src/tls/ffi.c

```diff
@@ -0,0 +1,428 @@
+#include <stdio.h>
+#include <memory.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/cdefs.h>
+#if __APPLE__ 
+#include <sys/errno.h> // OS/X only provides include/sys/errno.h
+#else
+#include <errno.h> // MinGW only provides include/errno.h
+#endif
+#include <caml/callback.h>
+#include <caml/alloc.h>
+#include <caml/memory.h>
+#include <caml/printexc.h>
+#include "mitlsffi.h"
+
+#define REDIRECT_STDIO 1
+
+#define MITLS_FFI_LIST \
+  MITLS_FFI_ENTRY(Config) \
+  MITLS_FFI_ENTRY(Connect) \
+  MITLS_FFI_ENTRY(Send) \
+  MITLS_FFI_ENTRY(Recv)
+ 
+// Pointers to ML code.  Initialized in FFI_mitls_init().  Invoke via caml_callback()
+#define MITLS_FFI_ENTRY(x) value* g_mitls_FFI_##x;
+MITLS_FFI_LIST
+#undef MITLS_FFI_ENTRY
+
+// Pass a C pointer into F* and recover it back.  OCaml limits integers to 2^30/2^62
+// so shift right by 1 before conversion to OCaml.  The low bit must be 0 in order to
+// meet structure alignment rules, so this is not lossy.
+_Static_assert(sizeof(size_t) <= sizeof(value), "OCaml value isn't large enough to hold a C pointer");
+#define PtrToValue(p) Val_long(((size_t)p)>>1)
+#define ValueToPtr(v) ((void*)((Long_val(v)<<1)))
+
+typedef struct mitls_state {
+    value fstar_state;    // a GC root representing an F*-side state object
+} mitls_state;
+
+// Support temporary redirection of stdout and stderr
+typedef struct {
+    char *out_name;
+    char *err_name;
+    int fd_out_old;
+    int fd_err_old;
+    int fd_out_new;
+    int fd_err_new;
+} mitls_redirect;
+
+//
+// Initialize miTLS.
+//
+//  Called once ahead of using miTLS
+//
+//  Returns:  0 for error, nonzero for success
+//
```

Comment:
I'm not sure what you're asking.  No, FFI can be called from non-C code and code that doesn't implement main().  For example, I expect it will work when called from C# code compiled into a DLL.  Do you see a problem I overlooked?

stdout/stderr redirection is a temporary workaround until a separate work item has been completed, to remove stdout/stderr usage from miTLS.